azurerm/resource_arm_iothub_shared_access_policy.go

```diff
@@ -0,0 +1,354 @@
+		if err != nil {
+			return fmt.Errorf("Error loading Shared Access Profiles of IotHub %q (Resource Group %q): %+v", iothubName, resourceGroup, err)
+		}
+		existingAccessPolicy := accessPolicyIterator.Value()
+
+		if strings.EqualFold(*existingAccessPolicy.KeyName, keyName) {
+			if d.IsNewResource() && requireResourcesToBeImported {
+				return tf.ImportAsExistsError("azurerm_iothub_shared_access_policy", resourceId)
+			}
+			accessPolicies = append(accessPolicies, expandedAccessPolicy)
+			alreadyExists = true
+		} else {
+			accessPolicies = append(accessPolicies, existingAccessPolicy)
+		}
+	}
+
+	if d.IsNewResource() {
+		accessPolicies = append(accessPolicies, expandedAccessPolicy)
+	} else if !alreadyExists {
+		return fmt.Errorf("Unable to find Shared Access Policy %q defined for IotHub %q (Resource Group %q)", keyName, iothubName, resourceGroup)
+	}
+
+	iothub.Properties.AuthorizationPolicies = &accessPolicies
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, iothubName, iothub, "")
+	if err != nil {
+		return fmt.Errorf("Error updating IotHub %q (Resource Group %q) with Shared Access Profile %q: %+v", iothubName, resourceGroup, keyName, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for IotHub %q (Resource Group %q) to finish updating Shared Access Profile %q: %+v", iothubName, resourceGroup, keyName, err)
+	}
+
+	d.SetId(resourceId)
+
+	return resourceArmIotHubSharedAccessPolicyRead(d, meta)
+}
+
+func resourceArmIotHubSharedAccessPolicyRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).iothubResourceClient
+	ctx := meta.(*ArmClient).StopContext
+
+	parsedIothubSAPId, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	resourceGroup := parsedIothubSAPId.ResourceGroup
+	iothubName := parsedIothubSAPId.Path["IotHubs"]
+	keyName := parsedIothubSAPId.Path["IotHubKeys"]
+
+	accessPolicy, err := client.GetKeysForKeyName(ctx, resourceGroup, iothubName, keyName)
+	if err != nil {
+		if utils.ResponseWasNotFound(accessPolicy.Response) {
+			log.Printf("[DEBUG] Shared Access Policy %q was not found on IotHub %q (Resource Group %q) - removing from state", keyName, iothubName, resourceGroup)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error loading IotHub Shared Access Policy %q (Resource Group %q): %+v", iothubName, resourceGroup, err)
+	}
+
+	iothub, err := client.Get(ctx, resourceGroup, iothubName)
+	if err != nil {
+		return fmt.Errorf("Error loading IotHub %q (Resource Group %q): %+v", iothubName, resourceGroup, err)
+	}
+
+	d.Set("name", keyName)
+	d.Set("iothub_name", iothubName)
+	d.Set("resource_group_name", resourceGroup)
+
+	d.Set("primary_key", accessPolicy.PrimaryKey)
+	d.Set("primary_connection_string", getSharedAccessPolicyConnectionString(*iothub.Properties.HostName, keyName, *accessPolicy.PrimaryKey))
```

Comment:
So we don't check for an error with primitives, but we do for complex objects (confusing right)

So could we check this here?
```suggestion
	if err := d.Set("primary_connection_string", getSharedAccessPolicyConnectionString(*iothub.Properties.HostName, keyName, *accessPolicy.PrimaryKey)), err != nil {
		return fmt.ErrorF("error setting `primary_connection_string `: %v")
	}
```

sorry i didn't point this out last review

Reply:
done